### ISSUE TITLE
NegativeBinomial strange nan results

### ISSUE THREAD

**Issue by denadai2** [statsmodels/statsmodels#3002]
Hello,
I have this problem with a simple input. I have nans. Why?

```
from statsmodels.discrete.discrete_model import NegativeBinomial
from statsmodels.tools.tools import add_constant
x = [[-1.188915  , -0.08435071, -0.28315054, -0.20893618, -0.032652  ,
   -0.78140215, -1.12412096, -0.58311838, -0.33658259, -1.00854482,
    0.5602917 ,  0.12364335, -0.82968265, -0.24782673,  0.44457328,
   -0.82314254, -0.92598275, -0.66887313,  0.64740755, -0.14991105,
    0.67945103, -0.59348105,  0.71163572,  0.13444227, -0.63383368,
    0.07663709, -0.02295347,  0.23192179,  0.04760662,  1.22048489,
    1.14182317,  0.8513495 ,  0.02823298,  0.35429832,  0.22688185,
    0.41231436,  0.16037734,  1.20634128,  0.86755685,  0.38649392,
    0.11345882, -0.0658492 , -0.29634195, -0.54431373, -0.96913148,
   -0.23758453, -0.79162154, -0.39481993, -0.80035955, -0.37641622,
   -0.15328634, -0.18048563, -0.03855409, -1.05323874, -0.72261557,
   -0.00954654, -0.10123889, -0.32509544, -0.18531117, -0.39373254,
   -0.2192062 , -1.1495039 ,  0.2817525 ,  0.01639138,  0.78588443,
   -0.78631091, -1.36556965, -0.96956438, -0.31468739,  0.50418202,
    0.74912992,  1.44975063, -1.07238488,  0.67882015,  0.2358319 ,
   -0.34497058, -1.08813671, -0.91021771, -1.09965065, -0.37273791,
    1.31962174, -0.96249557,  4.15090692,  1.33879663,  0.59627037,
   -0.52658515,  0.73927896,  2.36540989,  4.8531637 ,  1.00918715,
    1.56362837,  3.13907687,  0.47788941, -8.96773216, -0.64230965,
   -0.88324051, -0.33456883,  1.50497418, -1.05583618, -0.74833245,
    0.05349834, -0.52070831, -0.3226729 , -0.35403134, -0.97925836,
   -1.15301304,  0.02519611]

Y_pro = [  46,  314,  269,  217,  289,  735,  437,  259,  866,  353,  290,
    444,  302,  327,  487,  202,  523,  216,  749,  800,  948,  289,
    495,  304,  145,  250,  432,  163,  318,  237, 1051,  552,  404,
    260,  619,  297,  395,  403,  452,  435,  320,  129,  218,   76,
     18,  182,  187,  124,  151,  265,  191,   92,  165,    1,  160,
    161,  226,   49,  234,  135,  552,  243,  465,  227,  545,   75,
    155,   57,  164,  296,  191,  441,  104,  369,  598,  194,   96,
    410,   14,  390,  560,  257, 1255,  727,  183,  168, 1459,  688,
   1179,  718,  745, 1474,   69,  217,   39,  196,  280,  556,  244,
    307,  284,  354,  194,  234,  161,  125,   99]
X_pro = add_constant(x)
model = NegativeBinomial(Y_pro, X_pro)
model_fitted = model.fit(maxiter=1000)
print("BIC", model_fitted.bic)
print(model_fitted.summary())
```

And the results are:

```
Warning: Desired error not necessarily achieved due to precision loss.
     Current function value: nan
     Iterations: 0
     Function evaluations: 1
     Gradient evaluations: 1
BIC nan
                 NegativeBinomial Regression Results                      
==============================================================================
Dep. Variable:                      y   No. Observations:                  107
Model:               NegativeBinomial   Df Residuals:                      105
Method:                           MLE   Df Model:                            1
Date:                Sat, 11 Jun 2016   Pseudo R-squ.:                     nan
Time:                        19:16:54   Log-Likelihood:                    nan
converged:                      False   LL-Null:                       -728.99
                                    LLR p-value:                       nan
==============================================================================
                 coef    std err          z      P>|z|      [95.0% Conf. Int.]
------------------------------------------------------------------------------
const             nan        nan        nan        nan           nan       nan
x1                nan        nan        nan        nan           nan       nan
alpha          1.1052        nan        nan        nan           nan       nan
```

   ==============================================================================

```
/home/nadai/.local/lib/python3.4/site-packages/statsmodels/base/model.py:466: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
 "Check mle_retvals", ConvergenceWarning)
  /home/nadai/.local/lib/python3.4/site-packages/statsmodels/base/model.py:466: ConvergenceWarning:    Maximum Likelihood optimization failed to converge. Check mle_retvals
 "Check mle_retvals", ConvergenceWarning)
```

**Comment by josef-pkt** [comment]
Thanks for the full example.

Poisson and NegativeBinomial are a bit of a headache in some cases because the exp overflows. scipy's bfgs should become more robust in future releases.

right now the only thing I managed is to start the optimization with `method='nm'` and a prespecified start_params, (after playing a bit it seems that bfgs and lbfgs also work with these Poisson Null/constant-only start_params)

```
res_nm = model.fit(maxiter=1000, method='nm', start_params=[np.log(np.mean(Y_pro)), 0, 0])
res = model.fit(start_params=res_nm.params)
```

Poisson is used for the start_params, but Poisson also fails in this example. I guess we need better options to try other start_params if Poisson fails.

adding to #1649

**Comment by denadai2** [comment]
Thanks

**Comment by kshedden** [comment]
I was able to get better behavior with `statsmodels.GLM`, which can also fit negative binomial models.  But you need to use the gradient optimization approach, not IRLS which is the default.  See below for code.  Unfortunately `GLM` will not estimate `alpha` for you, but you can loop through a range of `alpha` values and compare BICs.

```
for alpha in np.linspace(0.01, 1, 10):
    model = sm.GLM(Y_pro, X_pro, family=sm.families.NegativeBinomial(alpha=alpha))
    model_fitted = model.fit(method='newton')
    print(alpha, model_fitted.bic)
    print(model_fitted.params)
```

**Comment by denadai2** [comment]
@kshedden the only optimization approach is IRLS. Ref: http://statsmodels.sourceforge.net/devel/generated/statsmodels.genmod.generalized_linear_model.GLM.fit.html#statsmodels.genmod.generalized_linear_model.GLM.fit

It seems there is a strange behaviour with GLM as well. Bic gets always lower as I increase alpha.

**Comment by kshedden** [comment]
I was working from the github master.  We implemented gradient optimization
for GLM about a year ago, but there has not been a statsmodels release in
that time.

I made a plot of residuals on fitted values for your data, and there is no
evidence that the variance increases with the mean (much less that the
variance increases faster than direct proportionality with the mean).  So
alpha = 0 may be the best fit for this data (but IIRC the best BIC is for
alpha slightly greater than zero).

Kerby

**Comment by josef-pkt** [comment]
It doesn't produce nans with using current master and scipy.__version__  '0.18.1'

was most likely also fixed by #3928 which added better start_params

```
Optimization terminated successfully.
         Current function value: 6.693316
         Iterations: 9
         Function evaluations: 12
         Gradient evaluations: 12
BIC 1446.38819545
                     NegativeBinomial Regression Results                      
==============================================================================
Dep. Variable:                      y   No. Observations:                  107
Model:               NegativeBinomial   Df Residuals:                      105
Method:                           MLE   Df Model:                            1
Date:                Tue, 03 Oct 2017   Pseudo R-squ.:                 0.01756
Time:                        19:02:41   Log-Likelihood:                -716.18
converged:                       True   LL-Null:                       -728.99
                                        LLR p-value:                 4.197e-07
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const          5.8192      0.068     85.201      0.000       5.685       5.953
x1             0.2024      0.034      6.024      0.000       0.137       0.268
alpha          0.4954      0.064      7.690      0.000       0.369       0.622
==============================================================================
```
